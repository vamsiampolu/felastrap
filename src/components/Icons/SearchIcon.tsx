import { ComponentPropsWithoutRef } from "react";

export type IconProps = ComponentPropsWithoutRef<"svg">;

export const SearchIcon = ({
  stroke,
  fill,
  className,
  ...props
}: IconProps) => {
  return (
    <svg
      xmlns="http://www.w3.org/2000/svg"
      width={20}
      height={20}
      className={className ?? "DocSearch-Search-Icon"}
      viewBox="0 0 32 32"
      {...props}
    >
      <path
        d="M14.386 14.386l4.0877 4.0877-4.0877-4.0877c-2.9418 2.9419-7.7115 2.9419-10.6533 0-2.9419-2.9418-2.9419-7.7115 0-10.6533 2.9418-2.9419 7.7115-2.9419 10.6533 0 2.9419 2.9418 2.9419 7.7115 0 10.6533z"
        stroke={stroke ?? "#000"}
        fill={fill ?? "none"}
        fillRule="evenodd"
        strokeLinecap="round"
        strokeLinejoin="round"
      />
    </svg>
  );
};
