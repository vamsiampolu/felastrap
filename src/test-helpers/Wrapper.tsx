import type { IRenderer } from "fela";
import { type PropsWithChildren } from "react";
import { RendererProvider, ThemeProvider } from "react-fela";

const Wrapper = ({
  renderer,
  theme = {},
  children,
}: PropsWithChildren<{ renderer: IRenderer; theme?: object }>) => (
  <ThemeProvider theme={theme}>
    <RendererProvider renderer={renderer}>{children}</RendererProvider>;
  </ThemeProvider>
);

export { Wrapper };
