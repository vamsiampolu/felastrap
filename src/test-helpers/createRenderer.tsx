import expandShorthand from "fela-plugin-expand-shorthand";
import sortClassNames from "fela-sort-classnames";
import sortMediaQueryMobileFirst from "fela-sort-media-query-mobile-first";
import prefixer from "fela-plugin-prefixer";
import placeholderPrefixer from "fela-plugin-placeholder-prefixer";
import { createRenderer as createFelaRenderer, type IRenderer } from "fela";
import { lightTheme } from "../reboot/lightTheme";

export const createRenderer = () => {
  const autoMergeLonghands = true;
  const enhancers = [sortClassNames(), sortMediaQueryMobileFirst()];
  const plugins = [
    prefixer(),
    placeholderPrefixer(),
    expandShorthand(autoMergeLonghands),
  ];

  return createFelaRenderer({
    plugins,
    enhancers,
  });
};

export const cssReboot = (renderer: IRenderer) => {
  renderer.renderStatic(lightTheme);
};
