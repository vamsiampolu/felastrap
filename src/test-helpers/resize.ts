export const resizeWidth = (width: number) => {
  window.innerWidth = width;
  dispatchEvent(new Event("resize"));
};

export const resizeHeight = (height: number) => {
  window.innerHeight = height;
  dispatchEvent(new Event("resize"));
};
