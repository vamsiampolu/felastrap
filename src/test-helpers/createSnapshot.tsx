import { IRenderer } from "fela";
import { renderToString } from "fela-tools";
import { ReactElement, ReactNode } from "react";

import { render, screen } from "@testing-library/react";
import { userEvent } from "@testing-library/user-event";
import {
  type UserEvent,
  type UserEventApi,
} from "@testing-library/user-event/dist/types/setup/setup";
import * as prettier from "prettier";
import { renderToStaticMarkup } from "react-dom/server";
import { Wrapper } from "./Wrapper";

async function formatCSS(css: string) {
  return prettier.format(css, {
    parser: "css",
    useTabs: false,
    tabWidth: 2,
  });
}

function formatHTML(str: string) {
  return prettier.format(str, { parser: "html" });
}

function renderToMarkup(component: ReactElement): string {
  return renderToStaticMarkup(component);
}

export const createSnapshot = async ({
  component,
  renderer,
  theme = {},
}: {
  component: ReactNode;
  renderer: IRenderer;
  theme?: object;
}) => {
  // reset renderer to have a clean setup
  renderer.clear();

  const element = (
    <Wrapper theme={theme} renderer={renderer}>
      {component}
    </Wrapper>
  );

  const view = renderToMarkup(element);
  const str = renderToString(renderer);

  const css = await formatCSS(str);
  const html = await formatHTML(view);

  return `${css}\n\n${html}`;
};

export type EventName = keyof UserEventApi | "focus" | "focus-visible";
export type CreateUserEventType = {
  eventName: EventName;
  component: ReactElement;
  renderer: IRenderer;
  testId: string;
  theme?: object;
  donotWrap?: boolean;
};

export type SetupEvent = {
  user: UserEvent;
  eventName: EventName;
  element: HTMLElement;
};
const setupEvent = async ({ user, eventName, element }: SetupEvent) => {
  switch (eventName) {
    case "hover": {
      await user.hover(element);
      break;
    }

    case "focus": {
      element.focus();
      break;
    }

    case "focus-visible": {
      element.contentEditable = "true";
      element.focus();
      element.contentEditable = "false";
      break;
    }
  }
};

const clearEvent = async ({ eventName, user, element }: SetupEvent) => {
  switch (eventName) {
    case "hover": {
      await user.unhover(element);
      break;
    }

    case "focus":
    case "focus-visible": {
      element.blur();
      break;
    }
  }
};

export const createUserEventSnapshot = async ({
  eventName,
  component,
  testId,
  renderer,
  theme = {},
  donotWrap = false,
}: CreateUserEventType) => {
  renderer.clear();

  const user = userEvent.setup();
  let wrappedComponent = component;
  if (!donotWrap) {
    wrappedComponent = (
      <Wrapper theme={theme} renderer={renderer}>
        {component}
      </Wrapper>
    );
  }

  render(wrappedComponent);

  const element = await screen.findByTestId(testId);

  setupEvent({ element, eventName, user });
  const view = await renderToMarkup(wrappedComponent);
  const utils = renderToString(renderer);
  const css = await formatCSS(utils);
  const html = await formatHTML(view);

  clearEvent({ eventName, element, user });

  return `${css}\n\n${html}`;
};
