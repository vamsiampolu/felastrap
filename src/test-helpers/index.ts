import { Wrapper } from "./Wrapper";
import { createRenderer, cssReboot } from "./createRenderer";
import { createSnapshot, createUserEventSnapshot } from "./createSnapshot";
import { resizeHeight, resizeWidth } from "./resize";

export {
  createRenderer,
  cssReboot,
  createSnapshot,
  createUserEventSnapshot,
  resizeWidth,
  resizeHeight,
  Wrapper,
};
