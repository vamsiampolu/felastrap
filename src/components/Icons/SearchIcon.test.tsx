import { SearchIcon } from "./SearchIcon";
import { IRenderer } from "fela";
import { createRenderer, createSnapshot, cssReboot } from "../../test-helpers";

describe("components/Icons/SearchIcon", () => {
  let renderer: IRenderer;
  beforeEach(() => {
    renderer = createRenderer();
    cssReboot(renderer);
  });

  afterEach(() => {
    renderer.clear();
  });

  it("should render the search icon", async () => {
    const component = <SearchIcon />;
    const snapshot = await createSnapshot({ component, renderer });
    expect(snapshot).toMatchSnapshot();
  });

  it("should render the search icon with a className", async () => {
    const component = <SearchIcon className="search" />;
    const snapshot = await createSnapshot({ component, renderer });
    expect(snapshot).toMatchSnapshot();
  });
});
