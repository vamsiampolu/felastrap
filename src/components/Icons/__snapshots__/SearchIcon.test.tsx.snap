// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`components/Icons/SearchIcon should render the search icon 1`] = `
"

<svg
  xmlns="http://www.w3.org/2000/svg"
  width="20"
  height="20"
  class="DocSearch-Search-Icon"
  viewBox="0 0 32 32"
>
  <path
    d="M14.386 14.386l4.0877 4.0877-4.0877-4.0877c-2.9418 2.9419-7.7115 2.9419-10.6533 0-2.9419-2.9418-2.9419-7.7115 0-10.6533 2.9418-2.9419 7.7115-2.9419 10.6533 0 2.9419 2.9418 2.9419 7.7115 0 10.6533z"
    stroke="#000"
    fill="none"
    fill-rule="evenodd"
    stroke-linecap="round"
    stroke-linejoin="round"
  ></path></svg
>;
"
`;

exports[`components/Icons/SearchIcon should render the search icon with a className 1`] = `
"

<svg
  xmlns="http://www.w3.org/2000/svg"
  width="20"
  height="20"
  class="search"
  viewBox="0 0 32 32"
>
  <path
    d="M14.386 14.386l4.0877 4.0877-4.0877-4.0877c-2.9418 2.9419-7.7115 2.9419-10.6533 0-2.9419-2.9418-2.9419-7.7115 0-10.6533 2.9418-2.9419 7.7115-2.9419 10.6533 0 2.9419 2.9418 2.9419 7.7115 0 10.6533z"
    stroke="#000"
    fill="none"
    fill-rule="evenodd"
    stroke-linecap="round"
    stroke-linejoin="round"
  ></path></svg
>;
"
`;
